consensus/state_processing/src/per_epoch_processing/altair/participation_cache.rs

```diff
@@ -0,0 +1,409 @@
+//! Provides the `ParticipationCache`, a custom Lighthouse cache which attempts to reduce CPU and
+//! memory usage by:
+//!
+//! - Caching a map of `validator_index -> participation_flags` for all active validators in the
+//!   previous and current epochs.
+//! - Caching the total balances of:
+//!   - All active validators.
+//!   - All active validators matching each of the three "timely" flags.
+//! - Caching the "eligible" validators.
+//!
+//! Additionally, this cache is returned from the `altair::process_epoch` function and can be used
+//! to get useful summaries about the validator participation in an epoch.
+
+use safe_arith::{ArithError, SafeArith};
+use std::collections::HashMap;
+use types::{
+    consts::altair::{
+        NUM_FLAG_INDICES, TIMELY_HEAD_FLAG_INDEX, TIMELY_SOURCE_FLAG_INDEX,
+        TIMELY_TARGET_FLAG_INDEX,
+    },
+    BeaconState, BeaconStateError, ChainSpec, Epoch, EthSpec, ParticipationFlags, RelativeEpoch,
+};
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    InvalidFlagIndex(usize),
+}
+
+/// A balance which will never be below the specified `minimum`.
+///
+/// This is an effort to ensure the `EFFECTIVE_BALANCE_INCREMENT` minimum is always respected.
+#[derive(PartialEq, Debug, Clone, Copy)]
+struct Balance {
+    raw: u64,
+    minimum: u64,
+}
+
+impl Balance {
+    /// Initialize the balance to `0`, or the given `minimum`.
+    pub fn zero(minimum: u64) -> Self {
+        Self { raw: 0, minimum }
+    }
+
+    /// Returns the balance with respect to the initialization `minimum`.
+    pub fn get(&self) -> u64 {
+        std::cmp::max(self.raw, self.minimum)
+    }
+
+    /// Add-assign to the balance.
+    pub fn safe_add_assign(&mut self, other: u64) -> Result<(), ArithError> {
+        self.raw.safe_add_assign(other)
+    }
+}
+
+/// Caches the participation values for one epoch (either the previous or current).
+#[derive(PartialEq, Debug)]
+struct SingleEpochParticipationCache {
+    /// Maps an active validator index to their participation flags.
+    ///
+    /// To reiterate, only active and unslashed validator indices are stored in this map.
+    ///
+    /// ## Note
+    ///
+    /// It would be ideal to maintain a reference to the `BeaconState` here rather than copying the
+    /// `ParticipationFlags`, however that would cause us to run into mutable reference limitations
+    /// upstream.
```

Comment:
_One day_ we can try a copy-on-write data structure for the participation flags, in which case duplicating it here will be as cheap as cloning an `Arc`.